Updates to service.nextpvr for v7

emveepee opened the pull request:
libmediainfo:  Currently libmediainfo only builds the static library.  Allow creation of shared library for applications that need it.  Tested with service.nextpvr, standalone mediainfo unchanged.

service.nextpvr: changes for v7 

- add libmediainfo.so and  hdhomerun_config
- add power management reset after sleep
- fix binary  not being marked +x  on manual updates

emveepee committed:
libmediainfo:  add shared library build

Currently libmediainfo only builds the static library.  Allow creation of shared library for applications that need it, including service.nextpvr Tested with service.nextpvr, standalone mediainfo unchanged.

heitbaum commented:
> Currently libmediainfo only builds the static library. Allow creation of shared library for applications that need it. Tested with service.nextpvr, standalone mediainfo unchanged.

Is the intent to use this in ${ADDON_BUILD}/${PKG_ADDON_ID}/lib.private and then add the LD_LIBRARY_PATH in packages/addons/service/nextpvr/source/bin/nextpvr.start or something like how we included the icu libraries in dotnet? https://github.com/LibreELEC/LibreELEC.tv/blob/0f7e211b8e2b7ffc2faf0a08474b56dad859e433/packages/addons/tools/dotnet-runtime/package.mk#L41

emveepee commented:
> Is the intent to use this in ${ADDON_BUILD}/${PKG_ADDON_ID}/lib.private and then add the LD_LIBRARY_PATH in packages/addons/service/nextpvr/source/bin/nextpvr.start or something like how we included the icu libraries in dotnet?

Basically my future plan is to make these changes to the service.nextpvr package (you can ignore libhdhomerun)  Others could use shared-lib as they wish

```
PKG_DEPENDS_TARGET="toolchain libhdhomerun libmediainfo"
...
mkdir -p ${INSTALL}/bin
cp $(get_install_dir libmediainfo)/shared-lib/libmediainfo.so ${INSTALL}/bin
cp $(get_build_dir libhdhomerun)/hdhomerun_config ${INSTALL}/bin
```

and this change to the nextpvr-downloader since nextpvr-bin is already in the dotnet LD_LIBRARY_PATH  The nextpvr-bin folder gets deleted with each install so I can't install to it directly

```
ln -s ${ADDON_DIR}/bin/libmediainfo.so ${ADDON_DIR}/nextpvr-bin
```

heitbaum commented:
> > Is the intent to use this in ${ADDON_BUILD}/${PKG_ADDON_ID}/lib.private and then add the LD_LIBRARY_PATH in packages/addons/service/nextpvr/source/bin/nextpvr.start or something like how we included the icu libraries in dotnet?
> 
> Basically my future plan is to make these changes to the service.nextpvr package (you can ignore libhdhomerun) Others could use shared-lib as they wish
> 
> ```
> PKG_DEPENDS_TARGET="toolchain libhdhomerun libmediainfo"
> ...
> mkdir -p ${INSTALL}/bin
> cp $(get_install_dir libmediainfo)/shared-lib/libmediainfo.so ${INSTALL}/bin
> cp $(get_build_dir libhdhomerun)/hdhomerun_config ${INSTALL}/bin
> ```
> 
> and this change to the nextpvr-downloader since nextpvr-bin is already in the dotnet LD_LIBRARY_PATH The nextpvr-bin folder gets deleted with each install so I can't install to it directly
> 
> ```
> ln -s ${ADDON_DIR}/bin/libmediainfo.so ${ADDON_DIR}/nextpvr-bin
> ```

I can't see where the nextpvr-bin is included in the LD_LIBRARY_PATH. There are a couple addons that have LD_LIBRARY_PATH set in them post the lib.private cleanup - example being `addons/service/tigervnc/source/bin/tigervnc.start`, also ttyd and chrome - this should address the local "copies issue". 

For the hdhomerun change - wouldn't a dependency on `driver.dvb.hdhomerun` be the way to go?

emveepee commented:
> I can't see where the nextpvr-bin is included in the LD_LIBRARY_PATH. There are a couple addons that have LD_LIBRARY_PATH set in them post the lib.private cleanup - example being `addons/service/tigervnc/source/bin/tigervnc.start`, also ttyd and chrome - this should address the local "copies issue".

Technically it does not need to be defined since it is the default for the netcore NextPVRService dependency manifest.  The service would need to set it if it was placed in a non standard location.  If you think that is a better approach for exposing the shared library that is fine.

> For the hdhomerun change - wouldn't a dependency on `driver.dvb.hdhomerun` be the way to go?

I think I tried that but the binary hdhomerun_config was not in the path.  Even then service.nextpvr doesn't need the whole driver and nothing needs to be configured so this seems less confusing.

heitbaum commented on packages/addons/addon-depends/multimedia-tools-depends/depends/libmediainfo/package.mk:
```diff
@@ -3,6 +3,7 @@
 
 PKG_NAME="libmediainfo"
 PKG_VERSION="24.06"
+PKG_REV="1"
```
unnecessary change

emveepee replied:
That was too track any future changes since I could not find a model for creating .so files  I will remove it.

heitbaum commented on packages/addons/addon-depends/multimedia-tools-depends/depends/libmediainfo/package.mk:
```diff
@@ -38,4 +39,8 @@ makeinstall_target() {
   done
   cp -P .libs/* ${INSTALL}/usr/lib
   cp -P libmediainfo.pc ${INSTALL}/usr/lib/pkgconfig
+
```
remove this block as we can simplify the copy to only include the .a (as the .la is a broken symlink)

heitbaum commented on packages/addons/addon-depends/multimedia-tools-depends/depends/libmediainfo/package.mk:
```diff
@@ -38,4 +39,8 @@ makeinstall_target() {
   done
   cp -P .libs/* ${INSTALL}/usr/lib
```
change `cp -P .libs/* ${INSTALL}/usr/lib` to
`cp -P .libs/libmediainfo.a ${INSTALL}/usr/lib`

Suggest making sure the commit mentions that the .so file is not copied in (but used by nextpvr)

heitbaum reviewed (changes requested):
would like to see the nextpvr commit included in the PR. Would like to see the .so file located in a location that was not bin or lib - as the addon is not sharing this with other addons / e.g. lib.private

heitbaum commented:
> > For the hdhomerun change - wouldn't a dependency on `driver.dvb.hdhomerun` be the way to go?
> 
> I think I tried that but the binary hdhomerun_config was not in the path. Even then service.nextpvr doesn't need the whole driver and nothing needs to be configured so this seems less confusing.

I had a loom at this - and probably the right way to go is include this `hdhomerun_config` binary in the addon - but I would suggest it now be in the "LE Path" - so not `/storage/.kodi/addons/yyy/bin`

emveepee commented:
> I had a look at this - and probably the right way to go is include this `hdhomerun_config` binary in the addon - but I would suggest it not be in the "LE Path" - so not `/storage/.kodi/addons/yyy/bin`

Can you explain way it shouldn't be in the path?  When the package is installed on Debian based and other Linux distros like Arch it is in /usr/bin.  Because of the nature of the LE having to set the Environment in system.d seems risky to me.

emveepee commented:
> would like to see the nextpvr commit included in the PR. Would like to see the .so file located in a location that was not bin or lib - as the addon is not sharing this with other addons / e.g. lib.private

Ok the  v7.0 version of service.nextpvr  (netcore 8) is still being finalized.  The libmedianfo.so changes do need to be final to be able to create a good package so I was treating it as an independent package.  Other package could use it.  

I will commit a package 6.1.5  for review but the release should be on hold until v7.0.0 is released to github, only the version number changes.

heitbaum commented on packages/addons/service/nextpvr/package.mk:
```diff
@@ -25,4 +25,9 @@ addon() {
 
 post_install_addon() {
   sed -e "s/@NEXTPVR_VERSION@/${PKG_ADDON_VERSION}/g" -i "${INSTALL}/bin/nextpvr-downloader"
+
+  mkdir -p ${INSTALL}/bin
+  cp $(get_install_dir libmediainfo)/shared-lib/libmediainfo.so ${INSTALL}/bin
+  cp $(get_build_dir libhdhomerun)/hdhomerun_config ${INSTALL}/bin
+
```
.so file should not be in LE PATH. I also don't think hdhomerun_config should be added to the LE PATH, as this addon is nextpvr. Place it in the nextpvr-bin directory or in an lbin directory?
```suggestion

  mkdir -p ${INSTALL}/lib.private
  cp $(get_build_dir libmediainfo)/.libs/libmediainfo.so ${INSTALL}/lib.private
  cp $(get_build_dir libhdhomerun)/hdhomerun_config ${INSTALL}/bin

```

emveepee replied:
I actually thought nextpvr-bin was in the $PATH but I see I misread that /and I will be able to move the library from bin.

emveepee replied:
OK I had to make a slight change since .libs is not directly under the build directory.

heitbaum commented on packages/addons/service/nextpvr/source/bin/nextpvr-downloader:
```diff
@@ -61,6 +61,9 @@ fi
 kodi-send --action="Notification(Extracting NextPVR,starting,1000,${ICON})" >/dev/null
 mkdir -p ${ADDON_DIR}/nextpvr-bin
+
+ln -s ${ADDON_DIR}/bin/libmediainfo.so ${ADDON_DIR}/nextpvr-bin
```
I'm not in favour of this - but if that's how nextpvr works - and if grabs the .so file from nextpvr-bin - then ok. But should reference the lib.private directory. hdhomerun_config should be done similarly - it could be in bin.private and linked into nextpvr-bin. Both the .so file and the hdhomerun_config should not be added to the LE PATH.

emveepee replied:
It is dotnet doing a specific library search (it even checks the dotnet library folder) but the executable search don't work the same way.  I will check with the author if the default folder set be LE https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/addons/service/nextpvr/source/bin/nextpvr.start#L25 means it will be reliably found.

emveepee replied:
Actually it looks like the current working directory will be set ok  so I can probably use the symbolic link from bin.private so I will set that up tomorrow.  Thanks.

```

[code]
2024-09-26 23:44:25.913	[ERROR][26]	Unexpected error doing legacy HDHR tuning: System.ComponentModel.Win32Exception (2): An error occurred trying to start process 'hdhomerun_config' with working directory '/storage/.kodi/addons/service.nextpvr/nextpvr-bin'. No such file or directory
[/code]
```

heitbaum commented on packages/addons/service/nextpvr/source/addon.py:
```diff
@@ -75,6 +75,8 @@ def updateNextPVR(self):
             xbmc.log('Running: %s' % command, xbmc.LOGDEBUG)
             os.system(command)
             os.remove(archive)
+            command = 'find {0}/DeviceHost -name DeviceHostLinux -exec chmod 755 {{}} \;'.format(dest_folder)
```
Shouldn't this be in the nextpvr-downloader, and be a fully qualified path not variable driven?

emveepee replied:
There are two ways of downloading, the default in nextpvr-downloader does download a specific version based, this second manual download in python gets the current Linux build, if a user might want to update LE11 or download a new version before it is packaged for LE 12/13.  The generic build has no version attributes.

emveepee committed:
service.nextpvr: changes for v7

- download version 7.0.0
- add libmediainfo.so, comskip and  hdhomerun_config
- add power management reset after sleep
- fix binary  not being marked +x  on manual updates

emveepee commented on packages/addons/addon-depends/multimedia-tools-depends/depends/libmediainfo/package.mk:
```diff
@@ -36,6 +36,9 @@ makeinstall_target() {
+  
+  # move shared lib so mediainfo will not detect it and perform a static build
+  mkdir -p ${INSTALL}/shared-lib
+  cp -P .libs/libmediainfo.so* ${INSTALL}/shared-lib
```
OK done, this builds mediainfo correctly

emveepee commented:
@heitbaum what am I doing wrong the zip file that gets created for service.next had hdhomerun_config +x but  when installed they lose the attribute.
```
ls /storage/.kodi/addons/service.nextpvr/local/ -lt
total 7708
-rw-r--r--    1 root     root         68296 Sep 27 11:13 hdhomerun_config
-rw-r--r--    1 root     root       7821936 Sep 27 11:13 libmediainfo.so
```

heitbaum commented:
> @heitbaum what am I doing wrong the zip file that gets created for service.nextpvr has hdhomerun_config +x but when installed they lose the attribute.
> 
> ```
> ls /storage/.kodi/addons/service.nextpvr/local/ -lt
> total 7708
> -rw-r--r--    1 root     root         68296 Sep 27 11:13 hdhomerun_config
> -rw-r--r--    1 root     root       7821936 Sep 27 11:13 libmediainfo.so
> ```

Kodi unzip does carry the execute bit across. Have a look here. 

git grep chmod packages/addons/

emveepee commented:
> Kodi unzip does carry the execute bit across. Have a look here.
> 
When I unzip the same zip manually on the same box it is ok so the zip file and unzip do preserver the attribute.

```
unzip /storage/downloads/service.nextpvr-6.1.5~5.zip -O /tmp

ls -lt /tmp/service.nextpvr/local/
total 7708
-rwxr-xr-x    1 root     root         68296 Sep 27 19:42 hdhomerun_config
-rwxr-xr-x    1 root     root       7821936 Sep 27 19:42 libmediainfo.so
```

Yesterday I am pretty sure it worked to $(INSTALL)/bin  I will do both and compare.  Could this be a CoreELEC issue? (that is where I test)

emveepee commented:
It is odd, everything in bin is being marked +x 

Here is what the source zip has

```
CoreELEC1:/tmp # ls -lt service.nextpvr/bin/
total 76
-rwxr-xr-x    1 root     root         68296 Sep 27 20:00 hdhomerun_config
-rw-r--r--    1 root     root          3164 Sep 27 20:00 nextpvr-downloader
-rw-r--r--    1 root     root           664 Sep 27 20:00 nextpvr.start
and
CoreELEC1:/tmp # ls -lt service.nextpvr/bin/
total 76
-rwxr-xr-x    1 root     root         68296 Sep 27 20:00 hdhomerun_config
-rw-r--r--    1 root     root          3164 Sep 27 20:00 nextpvr-downloader
-rw-r--r--    1 root     root           664 Sep 27 20:00 nextpvr.start
```
but only /bin is +x including files that didn't start with that.
```
CoreELEC1:/tmp # ls -lt /storage/.kodi/addons/service.nextpvr/bin/
total 76
-rwxr-xr-x    1 root     root         68296 Sep 27 20:02 hdhomerun_config
-rwxr-xr-x    1 root     root          3164 Sep 27 20:02 nextpvr-downloader
-rwxr-xr-x    1 root     root           664 Sep 27 20:02 nextpvr.start

and nothing in local

CoreELEC1:/tmp # ls -lt /storage/.kodi/addons/service.nextpvr/local/
total 7708
-rw-r--r--    1 root     root         68296 Sep 27 20:02 hdhomerun_config
-rw-r--r--    1 root     root       7821936 Sep 27 20:02 libmediainfo.so

```

heitbaum commented:
> Testing with it copied twice It is odd, everything in bin is being marked +x
> 
> Here is what the source zip has
> 
> ```
> CoreELEC1:/tmp # ls -lt service.nextpvr/bin/
> total 76
> -rwxr-xr-x    1 root     root         68296 Sep 27 20:00 hdhomerun_config
> -rw-r--r--    1 root     root          3164 Sep 27 20:00 nextpvr-downloader
> -rw-r--r--    1 root     root           664 Sep 27 20:00 nextpvr.start
> and
> CoreELEC1:/tmp # ls -lt service.nextpvr/local/
> total 7708
> -rwxr-xr-x    1 root     root         68296 Sep 27 20:00 hdhomerun_config
> -rwxr-xr-x    1 root     root       7821936 Sep 27 20:00 libmediainfo.so
> ```
> 
> but only /bin is +x including files that didn't start with that.
> 
> ```
> CoreELEC1:/tmp # ls -lt /storage/.kodi/addons/service.nextpvr/bin/
> total 76
> -rwxr-xr-x    1 root     root         68296 Sep 27 20:02 hdhomerun_config
> -rwxr-xr-x    1 root     root          3164 Sep 27 20:02 nextpvr-downloader
> -rwxr-xr-x    1 root     root           664 Sep 27 20:02 nextpvr.start
> 
> and nothing in local
> 
> CoreELEC1:/tmp # ls -lt /storage/.kodi/addons/service.nextpvr/local/
> total 7708
> -rw-r--r--    1 root     root         68296 Sep 27 20:02 hdhomerun_config
> -rw-r--r--    1 root     root       7821936 Sep 27 20:02 libmediainfo.so
> ```

Your testing is correct.
- The zip files are correct.
- Kodi unzip doesn’t honour the x
- LE adds x to all /bin/

emveepee commented:
> Your testing is correct.
> 
>     * The zip files are correct.
>     * Kodi unzip doesn’t honour the x
>     * LE adds x to all /bin/

I have updated the PR with a chmod.  I would prefer if hdhomerun_config was in the path, but I can understand you having conflict issues if another addon did the same and created conflicts.

vpeter4 commented:
Which other addon has binaries also in `local` folder?

heitbaum commented:
> > Your testing is correct.
> > ```
> > * The zip files are correct.
> > * Kodi unzip doesn’t honour the x
> > * LE adds x to all /bin/
> > ```
> 
> I have updated the PR with a chmod. I would prefer if hdhomerun_config was in the path, but I can understand you having conflict issues if another addon did the same and created conflicts.

Looking good. All looks self contained 👍. Ready for merging?

emveepee commented:
> Looking good. All looks self contained 👍. Ready for merging?

No I still need to add the release version for v7 plus I also have to meet a request for some users for comskip, but I was using hdhomerun_config as a model before doing that.  Initially I thought I could simply install the package like tvheadedend since it is allowed to put comskip and .so modules in bin.  For comskip it doesn't need to be in the path, it is all user scripted.

emveepee commented:
@heitbaum I have added comskip and also the updated to the new version on NextPVR 7.0.0.  I will squash the two commits after you review the changes.  One concern I have is using the libx265.so.212  but I basically got the code from TVHeadend which actually uses a different version.

heitbaum commented:
Looking good. I have done some tidy up on the .so (to be consistent) and done a build. 

Have a look at my suggested in https://github.com/heitbaum/LibreELEC.tv/commit/24aabc41189800a50db9e2afc08723b5d089136b

Think that we are going to have a pain in managing the libx265.so - buts let’s see.

emveepee commented:
> Think that we are going to have a pain in managing the libx265.so - buts let’s see.

@CvH  has this been an issue with tvheadend?

mglae commented:
> Also how do I make nextpvr.power executable? I do see it here

Not required, any *.power regular file executed with a shell.

heitbaum commented:
Do you want to squash the last 2 commits?

emveepee commented:
> Do you want to squash the last 2 commits?

OK should be one commit now.  Thanks @mglae I saw tvheadedn has it executable but could seem to get it to commit.
